Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp

```diff
@@ -315,13 +319,68 @@ class UserScriptMessageTest : public WebViewTest {
         return m_userScriptMessage;
     }
 
+    static void asyncScriptMessageReceived(WebKitUserContentManager* userContentManager, WebKitUserScriptMessage* userScriptMessage, UserScriptMessageTest* test)
+    {
+        g_signal_handlers_disconnect_by_func(userContentManager, reinterpret_cast<gpointer>(asyncScriptMessageReceived), test);
+        if (!test->m_waitForScriptRun)
+            g_main_loop_quit(test->m_mainLoop);
+
+        if (test->m_userScriptMessageErrorToProduce)
+            webkit_user_script_message_error(userScriptMessage, test->m_userScriptMessageErrorToProduce.get());
+        else {
+            GRefPtr<JSCValue> jsValue {
+                jsc_value_new_string(jsc_value_get_context(webkit_user_script_message_get_value(userScriptMessage)), "Inception")
+            };
+            webkit_user_script_message_reply(userScriptMessage, jsValue.get());
+        }
+    }
+
+    JSCValue* waitUntilPromiseResolved(const char* handlerName)
+    {
+        m_userScriptMessageReply = nullptr;
+        m_userScriptMessageErrorMessage.reset();
+
+        GUniquePtr<char> signalName(g_strdup_printf("script-request-received::%s", handlerName));
+        g_signal_connect(m_userContentManager.get(), signalName.get(), G_CALLBACK(asyncScriptMessageReceived), this);
+
+        g_main_loop_run(m_mainLoop);
+        g_assert_false(m_waitForScriptRun);
+
+        return m_userScriptMessageReply.get();
+    }
+
     static void runJavaScriptFinished(GObject*, GAsyncResult* result, UserScriptMessageTest* test)
     {
         g_assert_true(test->m_waitForScriptRun);
         test->m_waitForScriptRun = false;
         g_main_loop_quit(test->m_mainLoop);
     }
 
+    static void runAsyncJavaScriptFinished(GObject* webView, GAsyncResult* result, UserScriptMessageTest* test)
+    {
+        g_assert_true(test->m_waitForScriptRun);
+        test->m_waitForScriptRun = false;
+        g_main_loop_quit(test->m_mainLoop);
+
+        GUniqueOutPtr<GError> error;
+        if (auto* jsResult = webkit_web_view_run_javascript_in_world_finish(WEBKIT_WEB_VIEW(webView), result, &error.outPtr())) {
+            auto* value = webkit_javascript_result_get_js_value(jsResult);
+            auto* context = jsc_value_get_context(value);
+            auto* exception = jsc_context_get_exception(context);
+            if (test->m_userScriptMessageErrorToProduce) {
+                g_assert(exception);
+                test->m_userScriptMessageErrorMessage.reset(g_strdup(jsc_exception_get_message(exception)));
+                g_assert_cmpstr(test->m_userScriptMessageErrorMessage.get(), ==, test->m_userScriptMessageErrorToProduce->message);
+            } else {
+                g_assert(!exception);
+                test->m_userScriptMessageReply = value;
+            }
+            webkit_javascript_result_unref(jsResult);
+        } else {
+            g_printerr("js result err: %s\n", error->message);
```

Comment:
The test should always work correctly and this `else` block should never be reached. We can check that with:

```suggestion
            g_assert_not_reached();
```